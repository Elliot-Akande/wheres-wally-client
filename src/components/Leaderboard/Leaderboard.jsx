import PropTypes from "prop-types";
import formatTime from "../../utils/formatTime";
import {
  Heading,
  Item,
  List,
  Name,
  Score,
  Status,
  StyledLeaderboard,
} from "./styles";

function Leaderboard({ data, loading }) {
  const getContent = () => {
    if (loading) {
      return <Status>Loading...</Status>;
    }

    if (data.scores.length === 0) {
      return <Status>No scores recorded.</Status>;
    }

    return (
      <List>
        {data.scores.map((entry) => (
          <Item key={entry._id}>
            <Name>{entry.name}</Name>
            <Score>{formatTime(entry.score)}</Score>
          </Item>
        ))}
      </List>
    );
  };

  return (
    <StyledLeaderboard>
      <Heading>Leaderboard</Heading>
      {getContent()}
    </StyledLeaderboard>
  );
}

Leaderboard.propTypes = {
  data: PropTypes.shape({
    scores: PropTypes.arrayOf(
      PropTypes.shape({
        _id: PropTypes.string,
        name: PropTypes.string,
        score: PropTypes.number,
      })
    ),
  }),
  loading: PropTypes.bool,
};

export default Leaderboard;
