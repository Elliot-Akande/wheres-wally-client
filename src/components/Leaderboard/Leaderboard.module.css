.container {
  width: 260px;
  padding: 16px;

  background-color: var(--neutral-500);
  border-radius: 8px;
}

.heading {
  margin: 0;
  padding-bottom: 16px;
}

.status {
  margin: 0;
}

.list {
  margin: 0;
  padding: 8px;
  height: 180px;

  display: flex;
  flex-direction: column;
  gap: 4px;

  overflow-y: scroll;

  list-style: none;
  background-color: var(--neutral-400);
  border-radius: 4px;
}

.item {
  height: max-content;

  display: flex;
  justify-content: space-between;
  gap: 8px;

  border-bottom: 1px solid var(--neutral-700);
}

.name {
  flex: 1;
  white-space: nowrap;
  overflow-x: clip;
  text-overflow: ellipsis;
}

.score {
  text-align: end;
}
