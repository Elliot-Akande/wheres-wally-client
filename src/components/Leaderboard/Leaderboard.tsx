import formatTime from "../../utils/formatTime";
import {
  Heading,
  Item,
  List,
  Name,
  Score,
  Status,
  StyledLeaderboard,
} from "./styles";

interface LeaderboardProps {
  data: {
    scores: Array<{
      _id: string;
      name: string;
      score: number;
    }>;
  } | null;
  loading: boolean;
}

function Leaderboard({ data, loading }: LeaderboardProps) {
  const getContent = () => {
    if (loading) {
      return <Status>Loading...</Status>;
    }

    if (!data) {
      return <Status>Data not present.</Status>;
    }

    if (data.scores.length === 0) {
      return <Status>No scores recorded.</Status>;
    }

    return (
      <List>
        {data.scores.map((entry) => (
          <Item key={entry._id}>
            <Name>{entry.name}</Name>
            <Score>{formatTime(entry.score)}</Score>
          </Item>
        ))}
      </List>
    );
  };

  return (
    <StyledLeaderboard>
      <Heading>Leaderboard</Heading>
      {getContent()}
    </StyledLeaderboard>
  );
}

export default Leaderboard;
