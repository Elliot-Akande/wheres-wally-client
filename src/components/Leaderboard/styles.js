import { styled } from "styled-components";

export const Status = styled.p`
  margin: 0;
`;

export const StyledLeaderboard = styled.div`
  width: 260px;
  padding: 16px;

  background-color: var(--neutral-500);
  border-radius: 8px;
`;

export const Heading = styled.h2`
  margin: 0;
  padding-bottom: 16px;
`;

export const List = styled.ul`
  margin: 0;
  padding: 8px;
  height: 180px;

  display: flex;
  flex-direction: column;
  gap: 4px;

  overflow-y: scroll;

  list-style: none;
  background-color: var(--neutral-400);
  border-radius: 4px;
`;

export const Item = styled.li`
  height: max-content;

  display: flex;
  justify-content: space-between;
  gap: 8px;

  border-bottom: 1px solid var(--neutral-700);
`;

export const Name = styled.span`
  flex: 1;
  white-space: nowrap;
  overflow-x: clip;
  text-overflow: ellipsis;
`;

export const Score = styled.span`
  text-align: end;
`;
