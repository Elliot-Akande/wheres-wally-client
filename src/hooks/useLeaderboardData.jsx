import { useState, useEffect } from "react";

const useLeaderboardData = (levelNum) => {
  const [data, setData] = useState(null);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(null);

  useEffect(() => {
    setData([
      {
        id: 3,
        name: "Anon",
        time: 32,
      },
      {
        id: 1,
        name: "Big Gary",
        time: 100,
      },
      {
        id: 2,
        name: "Gordon",
        time: 140,
      },
    ]);

    setLoading(false);
  }, []);

  return { data, loading, error };
};

export default useLeaderboardData;
