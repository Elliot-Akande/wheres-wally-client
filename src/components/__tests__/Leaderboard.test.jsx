import { render } from "@testing-library/react";
import { vi } from "vitest";
import Leaderboard from "../Leaderboard/Leaderboard";
import useLeaderboardData from "../Leaderboard/useLeaderboardData";
import { describe, it, expect } from "vitest";

const mocks = vi.hoisted(() => {
  return { useLeaderboardData: vi.fn() };
});

vi.mock("/src/components/Leaderboard/useLeaderboardData.jsx", () => ({
  default: mocks.useLeaderboardData,
}));

const levelNum = 3;
const data = [
  {
    id: 3,
    name: "Anon",
    time: 32,
  },
  {
    id: 1,
    name: "Big Gary",
    time: 100,
  },
  {
    id: 2,
    name: "Gordon",
    time: 140,
  },
];
const loading = false;
const error = null;

describe("Leaderboard", () => {
  it("renders heading and list on valid leaderboard data", () => {
    mocks.useLeaderboardData.mockReturnValue({ data, loading, error });

    const { container } = render(<Leaderboard levelNum={levelNum} />);

    expect(useLeaderboardData).toBe(mocks.useLeaderboardData);
    expect(useLeaderboardData).toHaveBeenCalled();

    expect(container).toMatchSnapshot();
  });

  it("renders heading and loading message when loading data", () => {
    mocks.useLeaderboardData.mockReturnValue({ data, loading: true, error });

    const { container } = render(<Leaderboard levelNum={levelNum} />);

    expect(useLeaderboardData).toBe(mocks.useLeaderboardData);
    expect(useLeaderboardData).toHaveBeenCalled();

    expect(container).toMatchSnapshot();
  });

  it("renders heading and error message on data fetch error", () => {
    mocks.useLeaderboardData.mockReturnValue({
      data,
      loading,
      error: new Error(),
    });

    const { container } = render(<Leaderboard levelNum={levelNum} />);

    expect(useLeaderboardData).toBe(mocks.useLeaderboardData);
    expect(useLeaderboardData).toHaveBeenCalled();

    expect(container).toMatchSnapshot();
  });

  it("renders heading and empty scores message when no data after loading", () => {
    mocks.useLeaderboardData.mockReturnValue({
      data: null,
      loading,
      error,
    });

    const { container } = render(<Leaderboard levelNum={levelNum} />);

    expect(useLeaderboardData).toBe(mocks.useLeaderboardData);
    expect(useLeaderboardData).toHaveBeenCalled();

    expect(container).toMatchSnapshot();
  });
});
