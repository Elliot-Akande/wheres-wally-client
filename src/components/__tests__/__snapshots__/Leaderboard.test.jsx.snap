// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Leaderboard > renders heading and empty scores message when no data after loading 1`] = `
<div>
  <div
    class="_container_acb05c"
  >
    <h2
      class="_heading_acb05c"
    >
      Leaderboard
    </h2>
    <p
      class="_status_acb05c"
    >
      No scores recorded.
    </p>
  </div>
</div>
`;

exports[`Leaderboard > renders heading and list on valid leaderboard data 1`] = `
<div>
  <div
    class="_container_acb05c"
  >
    <h2
      class="_heading_acb05c"
    >
      Leaderboard
    </h2>
    <ul
      class="_list_acb05c"
    >
      <li
        class="_item_acb05c"
      >
        <div
          class="_name_acb05c"
        >
          Anon
        </div>
        <div
          class="_score_acb05c"
        >
          32
        </div>
      </li>
      <li
        class="_item_acb05c"
      >
        <div
          class="_name_acb05c"
        >
          Big Gary
        </div>
        <div
          class="_score_acb05c"
        >
          100
        </div>
      </li>
    </ul>
  </div>
</div>
`;

exports[`Leaderboard > renders heading and loading message when loading data 1`] = `
<div>
  <div
    class="_container_acb05c"
  >
    <h2
      class="_heading_acb05c"
    >
      Leaderboard
    </h2>
    <p
      class="_status_acb05c"
    >
      Loading...
    </p>
  </div>
</div>
`;
